### PR TITLE
Support upload trace in Zipkin Lens UI

### PR DESCRIPTION
**Pull request by tacigar**
Related: #2350 
History: #2424

### CODE REVIEW

**Comment by tacigar** [comment]
Oh, I forgot to remove traceId component from search bar...

**Comment by tacigar** [comment]
Removed TraceId component from search bar...
Please refer #2368 🙇

**Comment by tacigar** [comment]
**Search by trace ID**
traceId

**Comment by tacigar** [comment]
**Upload Trace JSON**
fromJson

**Comment by tacigar** [comment]
**How it works**
fromJson

**Comment by tacigar** [comment]
Current component architecture is a bit complex.
I drew the graph.

components

**Comment by codefromthecrypt** [comment]
thanks! please add the "how it works" image to the README at the bottom under a markdown section named design or a separate DESIGN.md

File: zipkin-lens/src/components/GlobalSearch/GlobalDropdownMenu.js [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+      const { result } = fileReader;
+      let rawTrace;
+      try {
+        rawTrace = JSON.parse(result);
+        setTrace(rawTrace);
+        history.push({
+          pathname: '/zipkin/traceViewer',
+        });
+      } catch (error) {
+        // Do nothing
```
Comment (codefromthecrypt) [inline zipkin-lens/src/components/GlobalSearch/GlobalDropdownMenu.js]: maybe alert the json is malformed? I assume this catch will result in someone uploading an incorrect file (ex just one span) and getting no feedback about it. correct?

**Review by codefromthecrypt — APPROVED** [review]
Looks good. I like the design and I feel this maximizes the UI functionality without clutter.

one small note about error handling.

File: zipkin-lens/src/components/GlobalSearch/GlobalDropdownMenu.js (hunk 1 above)
Comment (tacigar) [reply zipkin-lens/src/components/GlobalSearch/GlobalDropdownMenu.js]: Yes.
Do you think that it is better to display the alert dialog?

**Comment by codefromthecrypt** [comment]
note: the view trace screen blows up (white screen) if using v1 json

```
TypeError: "e is undefined"

    getServiceNameColor http://localhost:9411/zipkin/app-d005bdc780c99e8fec7a.min.js:12 value http://localhost:9411/zipkin/app-d005bdc780c99e8fec7a.min.js:88 value http://localhost:9411/zipkin/app-d005bdc780c99e8fec7a.min.js:88 
```

In the old uploadTrace.js we had something like this. If you import it, please also import the unit tests.

```javascript
export function ensureV2(trace) {
  if (!Array.isArray(trace) || trace.length === 0) {
    throw new Error('input is not a list');
  }
  const first = trace[0];
  if (!first.traceId || !first.id) {
    throw new Error('List<Span> implies at least traceId and id fields');
  }
  if (first.binaryAnnotations || (!first.localEndpoint && !first.remoteEndpoint && !first.tags)) {
    throw new Error(
      'v1 format is not supported. For help, contact https://gitter.im/openzipkin/zipkin');
  }
}
```

**Comment by codefromthecrypt** [comment]
Looks good besides the blowup. I tested uploading a v2 json and also get trace by ID

File: zipkin-lens/src/components/GlobalSearch/GlobalDropdownMenu.js (hunk 1 above)
Comment (codefromthecrypt) [reply zipkin-lens/src/components/GlobalSearch/GlobalDropdownMenu.js]: old UI did something like this..

Screen Shot 2019-03-19 at 9 25 51 AM

Comment (tacigar) [reply zipkin-lens/src/components/GlobalSearch/GlobalDropdownMenu.js]: OK, Thank you.
I'll do so

**Comment by tacigar** [comment]
> note: the view trace screen blows up (white screen) if using v1 json
> 
> ```
> TypeError: "e is undefined"
> 
>     getServiceNameColor http://localhost:9411/zipkin/app-d005bdc780c99e8fec7a.min.js:12 value http://localhost:9411/zipkin/app-d005bdc780c99e8fec7a.min.js:88 value http://localhost:9411/zipkin/app-d005bdc780c99e8fec7a.min.js:88 
> ```
> In the old uploadTrace.js we had something like this. If you import it, please also import the unit tests.
> 
> ```js
> export function ensureV2(trace) {
>   if (!Array.isArray(trace) || trace.length === 0) {
>     throw new Error('input is not a list');
>   }
>   const first = trace[0];
>   if (!first.traceId || !first.id) {
>     throw new Error('List<Span> implies at least traceId and id fields');
>   }
>   if (first.binaryAnnotations || (!first.localEndpoint && !first.remoteEndpoint && !first.tags)) {
>     throw new Error(
>       'v1 format is not supported. For help, contact https://gitter.im/openzipkin/zipkin');
>   }
> }
> ```

I see!
I'll do so

**Comment by tacigar** [comment]
**If uploaded file is not JSON...**

not-json

**If v1 format file is uploaded...**

v1-trace

**Comment by codefromthecrypt** [comment]
goodie!

**Review by codefromthecrypt — APPROVED** [review]

**Comment by shakuzen** [comment]
Great work, @tacigar!

**Pull request merged**